examples/simple_trainer_mip_splatting.py

```diff
@@ -0,0 +1,1069 @@
+            self.pose_perturb = CameraOptModule(len(self.trainset)).to(self.device)
+            self.pose_perturb.random_init(cfg.pose_noise)
+
+        self.app_optimizers = []
+        if cfg.app_opt:
+            self.app_module = AppearanceOptModule(
+                len(self.trainset), feature_dim, cfg.app_embed_dim, cfg.sh_degree
+            ).to(self.device)
+            # initialize the last layer to be zero so that the initial output is zero.
+            torch.nn.init.zeros_(self.app_module.color_head[-1].weight)
+            torch.nn.init.zeros_(self.app_module.color_head[-1].bias)
+            self.app_optimizers = [
+                torch.optim.Adam(
+                    self.app_module.embeds.parameters(),
+                    lr=cfg.app_opt_lr * math.sqrt(cfg.batch_size) * 10.0,
+                    weight_decay=cfg.app_opt_reg,
+                ),
+                torch.optim.Adam(
+                    self.app_module.color_head.parameters(),
+                    lr=cfg.app_opt_lr * math.sqrt(cfg.batch_size),
+                ),
+            ]
+
+        # Losses & Metrics.
+        self.ssim = StructuralSimilarityIndexMeasure(data_range=1.0).to(self.device)
+        self.psnr = PeakSignalNoiseRatio(data_range=1.0).to(self.device)
+        self.lpips = LearnedPerceptualImagePatchSimilarity(normalize=True).to(
+            self.device
+        )
+
+        # Viewer
+        if not self.cfg.disable_viewer:
+            self.server = viser.ViserServer(port=cfg.port, verbose=False)
+            self.viewer = nerfview.Viewer(
+                server=self.server,
+                render_fn=self._viewer_render_fn,
+                mode="training",
+            )
+
+        # Running stats for prunning & growing.
+        n_gauss = len(self.splats["means3d"])
+        self.running_stats = {
+            "grad2d": torch.zeros(n_gauss, device=self.device),  # norm of the gradient
+            "count": torch.zeros(n_gauss, device=self.device, dtype=torch.int),
+        }
+
+    def rasterize_splats(
+        self,
+        camtoworlds: Tensor,
+        Ks: Tensor,
+        width: int,
+        height: int,
+        **kwargs,
+    ) -> Tuple[Tensor, Tensor, Dict]:
+        means = self.splats["means3d"]  # [N, 3]
+        # quats = F.normalize(self.splats["quats"], dim=-1)  # [N, 4]
+        # rasterization does normalization internally
+        quats = self.splats["quats"]  # [N, 4]
+        scales = torch.exp(self.splats["scales"])  # [N, 3]
+        opacities = torch.sigmoid(self.splats["opacities"])  # [N,]
+        filters = self.splats["filters"]  # [N,]
+        
+        # apply 3D smoothing filter to scales and opacities
+        scales_square = torch.square(scales) # [N, 3]
+        det1 = scales_square.prod(dim=1)     # [N, ]
+        
+        scales_after_square = scales_square + torch.square(filters) [:, None] # [N, 1]
+        det2 = scales_after_square.prod(dim=1) # [N,]
+        coef = torch.sqrt(det1 / det2) # [N,]
```

Comment:
Could the scale be very small, sqrt is not differentiable near zero. Maybe consider add a clamp to avoid zero differentiation?